python/ray/_private/usage/usage_lib.py

```diff
@@ -124,3 +129,76 @@ def get_cluster_metadata(gcs_client, num_retries) -> dict:
             num_retries=num_retries,
         )
     )
+
+
+def generate_report_data(cluster_metadata: dict) -> dict:
+    """Generate the report data.
+
+    Note that we don't perform schema validation.
+
+    The server side is in charge of schema validation, and the schema
+    will be available via public Pypi wheel. The e2e validation will
+    be done from the test.
+
+    cluster_metadata (dict): The cluster metadata of the system generated by
+        `_generate_cluster_metadata`.
+    """
+    data = cluster_metadata.copy()
+    data["collect_timestamp_ms"] = int(time.time() * 1000)
+    return data
+
+
+def write_usage_data(data: dict, dir_path: str) -> None:
+    """Write the usage data to the directory.
+
+    Params:
+        data (dict): Data to report
+        dir_path (Path): The path to the directory to write usage data.
+
+    Raises:
+
+    """
+    # Atomically update the file.
+    dir_path = Path(dir_path)
+    destination = dir_path / usage_constant.USAGE_STATS_FILE
+    temp = dir_path / "usage_stats_tmp.json"
+    with temp.open(mode="w") as json_file:
+        json_file.write(json.dumps(data))
+    temp.rename(destination)
+
+
+def report_usage_data(url: str, data: dict) -> None:
+    """Report the usage data to the hard-coded usage server.
+
+    Params:
+        url (str): The URL to update resource usage.
+        data (dict): Data to report.
+
+    Raises:
+        requests.HTTPError if requests fails.
+    """
+    r = requests.request(
+        "POST",
+        url,
+        headers={"Content-Type": "application/json"},
+        json=data,
+        timeout=10,
+    )
+    r.raise_for_status()
+    return r
+
+
+async def write_usage_data_async(data: dict, dir_path: str, executor: Executor) -> None:
+    loop = asyncio.get_running_loop()
+    try:
+        await loop.run_in_executor(executor, write_usage_data, data, dir_path)
+    except Exception as e:
+        logger.exception(e)
```

Comment:
Maybe saying something like this exception can be totally ignored since it won't affect anything. I'd rather prefer info/warn here actually: operator shouldn't be panic when they see this log.